File: apprefurbished/src/main/java/com/androidcollider/easyfin/home/HomeMVP.java
package com.androidcollider.easyfin.home;

import androidx.annotation.Nullable;
import androidx.core.util.Pair;

import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.PieData;

import java.util.Map;

import io.reactivex.Flowable;

/**
 * @author Ihor Bilous
 */

public interface HomeMVP {

    interface Model {

        Flowable<Pair<Map<String, double[]>, Map<String, double[]>>> getBalanceAndStatistic(int statisticPosition);

        Flowable<Map<String, double[]>> getBalance();

        Flowable<Map<String, double[]>> getStatistic(int statisticPosition);
    }

    interface View {

        void setBalanceAndStatistic(Pair<Map<String, double[]>, Map<String, double[]>> mapPair);

        void updateBalanceAndStatistic(Pair<Map<String, double[]>, Map<String, double[]>> mapPair);

        void updateBalanceAndStatisticAfterDBImport(Pair<Map<String, double[]>, Map<String, double[]>> mapPair);

        void updateBalance(Map<String, double[]> map);

        void updateStatistic(Map<String, double[]> map);

        boolean isNeedToConvert();

        int getBalanceCurrencyPosition();
    }

    interface Presenter {

        void setView(@Nullable HomeMVP.View view);

        void loadBalanceAndStatistic(int statisticPosition);

        void updateBalanceAndStatistic(int statisticPosition);

        void updateBalanceAndStatisticAfterDBImport(int statisticPosition);

        void updateBalance();

        void updateStatistic(int statisticPosition);

        boolean isStatisticEmpty();

        String getFormattedBalance(double[] balance);

        String getFormattedStatistic();

        double[] getCurrentBalance(int position);

        void updateTransactionStatisticArray(int posCurrency);

        BarData getDataSetMainBalanceHorizontalBarChart(double[] balance);

        BarData getDataSetMainStatisticHorizontalBarChart();

        PieData getDataSetMainStatisticPieChart();

        void updateRates();
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/home/HomeMVP.kt
package com.androidcollider.easyfin.home

import com.github.mikephil.charting.charts.HorizontalBarChart
import com.github.mikephil.charting.charts.PieChart
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.data.PieData
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
interface HomeMVP {
    interface Model {
        fun getBalanceAndStatistic(statisticPosition: Int):
                Single<Pair<Map<String, DoubleArray>, Map<String, DoubleArray>>>

        val balance: Single<Map<String, DoubleArray>>
        fun getStatistic(statisticPosition: Int): Single<Map<String, DoubleArray>>
    }

    interface View {
        fun setBalanceAndStatistic()

        fun updateBalanceAndStatistic()

        fun updateBalanceAndStatisticAfterDBImport()

        fun updateBalance()
        fun updateStatistic()
        val isNeedToConvert: Boolean
        val balanceCurrencyPosition: Int
    }

    interface Presenter {
        fun setView(view: View?)
        fun loadBalanceAndStatistic(statisticPosition: Int)
        fun updateBalanceAndStatistic(statisticPosition: Int)
        fun updateBalanceAndStatisticAfterDBImport(statisticPosition: Int)
        fun updateBalance()
        fun updateStatistic(statisticPosition: Int)
        val isStatisticEmpty: Boolean
        fun isBalanceEmpty(balance: DoubleArray): Boolean
        fun getFormattedBalance(balance: DoubleArray): String
        val formattedStatistic: String
        fun getCurrentBalance(posCurrency: Int): DoubleArray
        fun updateTransactionStatisticArray(posCurrency: Int)
        fun getDataSetMainBalanceHorizontalBarChart(
            balance: DoubleArray,
            chart: HorizontalBarChart
        ): BarData

        fun getDataSetMainStatisticHorizontalBarChart(chart: HorizontalBarChart): BarData
        fun getDataSetMainStatisticPieChart(chart: PieChart): PieData
        fun updateRates()
    }
}